source/simulator/core.cc

```diff
@@ -2377,95 +2375,32 @@ namespace aspect
         // then do the core work: assemble systems and solve
         solve_timestep ();
 
-        pcout << std::endl;
-
         // get new time step size
-        // for now the bool (convection/conduction dominated)
-        // returned by compute_time_step is unused, will be
-        // added to statistics later
-        const double new_time_step = termination_manager.check_for_last_time_step(
-                                       std::min(compute_time_step().first,
-                                                parameters.maximum_time_step));
+        const double new_time_step = compute_time_step();
 
         if (parameters.convert_to_years == true)
           statistics.add_value("Time step size (years)", new_time_step / year_in_seconds);
         else
           statistics.add_value("Time step size (seconds)", new_time_step);
 
-
-        // see if we have to start over with a new refinement cycle
+        // see if we have to start over with a new adaptive refinement cycle
         // at the beginning of the simulation
-        if ((timestep_number == 0) &&
-            (pre_refinement_step < parameters.initial_adaptive_refinement))
-          {
-            if (parameters.timing_output_frequency ==0)
-              computing_timer.print_summary ();
-
-            output_statistics();
-
-            if (parameters.run_postprocessors_on_initial_refinement)
-              postprocess ();
-
-            refine_mesh (max_refinement_level);
-            ++pre_refinement_step;
-            goto start_time_iteration;
-          }
-
-        // invalidate the value of pre_refinement_step since it will no longer be used from here on
-        if ( timestep_number == 0 )
-          pre_refinement_step = std::numeric_limits<unsigned int>::max();
+        if ((timestep_number == 0) && (maybe_do_initial_refinement(max_refinement_level)))
+          goto start_time_iteration;
```

Comment:
I find function calls in conditionals difficult to read and understand if they have side effects. I think it may be easier to read the code if it read
```
  if (timestep_number == 0)
  {
    const bool initial_refinement_done = maybe_do...(...);
    if (initial_refinement_done == true)
      goto ...;
  }
```
That's more code but may be easier to read.